Mathlib/CategoryTheory/Galois/Topology.lean

```diff
@@ -0,0 +1,145 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.CategoryTheory.Endomorphism
+import Mathlib.CategoryTheory.FintypeCat
+import Mathlib.Topology.Algebra.Group.Basic
+
+/-!
+
+# Topology of fundamental group
+
+In this file we define a natural topology on the automorphism group of a functor
+`F : C ⥤ FintypeCat`: It is defined as the subspace topology induced by the natural
+embedding of `Aut F` into `∀ X, Aut (F.obj X)` where
+`Aut (F.obj X)` carries the discrete topology.
+
+## References
+
+- Stacks Project: Tag 0BMQ
+
+-/
+universe u₁ u₂ v₁ v₂ v w
+
+namespace CategoryTheory
+
+namespace PreGaloisCategory
+
+open Functor
+
+variable {C : Type u₁} [Category.{u₂} C] (F : C ⥤ FintypeCat.{w})
+
+/-- For a functor `F : C ⥤ FintypeCat`, the canonical embedding of `Aut F` into
+the product over `Aut (F.obj X)` for all objects `X`. -/
+def autEmbedding : Aut F →* ∀ X, Aut (F.obj X) :=
+  MonoidHom.mk' (fun σ X ↦ σ.app X) (fun _ _ ↦ rfl)
+
+@[simp]
+lemma autEmbedding_apply (σ : Aut F) (X : C) : autEmbedding F σ X = σ.app X :=
+  rfl
+
+lemma autEmbedding_injective : Function.Injective (autEmbedding F) := by
+  intro σ τ h
+  ext X x
+  have : σ.app X = τ.app X := congr_fun h X
+  rw [← Iso.app_hom, ← Iso.app_hom, this]
+
+/-- We put the discrete topology on `F.obj X`. -/
+scoped instance (X : C) : TopologicalSpace (F.obj X) := ⊥
+
+@[scoped instance]
+lemma obj_discreteTopology (X : C) : DiscreteTopology (F.obj X) := ⟨rfl⟩
+
+/-- We put the discrete topology on `Aut (F.obj X)`. -/
+scoped instance (X : C) : TopologicalSpace (Aut (F.obj X)) := ⊥
+
+@[scoped instance]
+lemma aut_discreteTopology (X : C) : DiscreteTopology (Aut (F.obj X)) := ⟨rfl⟩
+
+/-- `Aut F` is equipped with the by the embedding into `∀ X, Aut (F.obj X)` induced embedding. -/
+instance : TopologicalSpace (Aut F) :=
+  TopologicalSpace.induced (autEmbedding F) inferInstance
+
+/-- The image of `Aut F` in `∀ X, Aut (F.obj X)` are precisely the compatible families of
+automorphisms. -/
+lemma autEmbedding_range :
+    Set.range (autEmbedding F) =
+      { a | ∀ (X Y : C) (f : X ⟶ Y), F.map f ≫ (a Y).hom = (a X).hom ≫ F.map f } := by
+  ext a
+  simp only [Set.mem_range, Set.mem_setOf_eq]
+  constructor
+  · intro ⟨σ, h⟩
+    rw [← h]
+    exact σ.hom.naturality
+  · intro h
+    use NatIso.ofComponents (fun X => a X)
+    rfl
+
+/-- The image of `Aut F` in `∀ X, Aut (F.obj X)` is closed. -/
+lemma autEmbedding_range_isClosed : IsClosed (Set.range (autEmbedding F)) := by
```

Comment:
I am not very much expert in the mathlib implementation of topology, but I would suggest using a slightly different skeleton of proof:
```lean
lemma autEmbedding_range :
    Set.range (autEmbedding F) =
      ⋂ (f : Arrow C), { a | F.map f.hom ≫ (a f.right).hom = (a f.left).hom ≫ F.map f.hom } := by
  sorry

-- to be moved
lemma isClosed_equalizer {X Y : Type*}  [TopologicalSpace X] [TopologicalSpace Y] (f g : X → Y)
    (hf : Continuous f) (hg : Continuous g) [T2Space Y]:
    IsClosed { x : X | f x = g x } := sorry

/-- The image of `Aut F` in `∀ X, Aut (F.obj X)` is closed. -/
lemma autEmbedding_range_isClosed : IsClosed (Set.range (autEmbedding F)) := by
  rw [autEmbedding_range]
  refine isClosed_iInter
    (fun f ↦ isClosed_equalizer (Y := F.obj f.left → F.obj f.right) _ _ ?_ ?_)
  · sorry
  · sorry
```

Reply:
Thanks! That's very nice and `isClosed_equalizer` actually already exists as `isClosed_eq`.